File: packages/typescript/src/network/http-client.ts
import type { DedClientConfig, ApiErrorResponse } from './types';
import type { X402Signer } from './x402';
import { parsePaymentRequired, buildPaymentHeader } from './x402';

/** Error thrown by DedHttpClient on API failures */
export class DedApiError extends Error {
  constructor(
    message: string,
    public readonly status: number,
    public readonly body?: ApiErrorResponse
  ) {
    super(message);
    this.name = 'DedApiError';
  }
}

/**
 * Standalone HTTP client for the DED Ingestion API.
 *
 * Uses native `fetch` (available in Node 18+ and all modern browsers).
 * Supports API key authentication or x402 pay-per-request via an {@link X402Signer}.
 */
export class DedHttpClient {
  private readonly baseUrl: string;
  private readonly apiKey: string | undefined;
  private readonly signer: X402Signer | undefined;
  private readonly timeout: number;

  constructor(config: DedClientConfig) {
    // Strip trailing slash for consistent URL joining
    this.baseUrl = config.baseUrl.replace(/\/+$/, '');
    this.apiKey = config.apiKey;
    this.signer = config.signer;
    this.timeout = config.timeout ?? 30_000;
  }

  /** GET request (authenticated with API key) */
  async get<T>(path: string, query?: Record<string, string>): Promise<T> {
    const url = this.buildUrl(path, query);
    return this.request<T>(url, { method: 'GET' });
  }

  /** GET request (public, no API key) */
  async getPublic<T>(path: string, query?: Record<string, string>): Promise<T> {
    const url = this.buildUrl(path, query);
    return this.request<T>(url, { method: 'GET', public: true });
  }

  /** POST request (authenticated with API key) */
  async post<T>(path: string, body: unknown): Promise<T> {
    const url = this.buildUrl(path);
    return this.request<T>(url, {
      method: 'POST',
      body: JSON.stringify(body),
      contentType: 'application/json',
    });
  }

  /** POST multipart request (authenticated with API key) */
  async postMultipart<T>(path: string, formData: FormData): Promise<T> {
    const url = this.buildUrl(path);
    return this.request<T>(url, {
      method: 'POST',
      formData,
    });
  }

  /** POST raw multipart body with explicit Content-Type (authenticated with API key) */
  async postRawMultipart<T>(path: string, body: Uint8Array, contentType: string): Promise<T> {
    const url = this.buildUrl(path);
    return this.request<T>(url, {
      method: 'POST',
      rawBody: body,
      contentType,
    });
  }

  private buildUrl(path: string, query?: Record<string, string>): string {
    const url = new URL(`${this.baseUrl}${path}`);
    if (query) {
      for (const [key, value] of Object.entries(query)) {
        if (value !== undefined && value !== '') {
          url.searchParams.set(key, value);
        }
      }
    }
    return url.toString();
  }

  private async request<T>(
    url: string,
    options: {
      method: string;
      body?: string;
      contentType?: string;
      rawBody?: Uint8Array;
      formData?: FormData;
      public?: boolean;
    }
  ): Promise<T> {
    const headers: Record<string, string> = {};

    if (!options.public && this.apiKey) {
      headers['X-Api-Key'] = this.apiKey;
    }

    if (options.contentType) {
      headers['Content-Type'] = options.contentType;
    }

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);

    try {
      const response = await fetch(url, {
        method: options.method,
        headers,
        body: options.rawBody ?? options.formData ?? options.body,
        signal: controller.signal,
      });

      // x402 payment flow: if 402 and we have a signer (but no API key), handle payment
      if (response.status === 402 && this.signer && !this.apiKey) {
        return this.handlePaymentRequired<T>(response, url, options);
      }

      if (!response.ok) {
        let errorBody: ApiErrorResponse | undefined;
        try {
          errorBody = (await response.json()) as ApiErrorResponse;
        } catch {
          // Response body may not be JSON
        }
        const message =
          errorBody?.message ??
          errorBody?.errors?.[0]?.message ??
          `HTTP ${response.status}: ${response.statusText}`;
        throw new DedApiError(message, response.status, errorBody);
      }

      return (await response.json()) as T;
    } finally {
      clearTimeout(timeoutId);
    }
  }

  private async handlePaymentRequired<T>(
    response: Response,
    url: string,
    options: {
      method: string;
      body?: string;
      contentType?: string;
      rawBody?: Uint8Array;
      formData?: FormData;
    }
  ): Promise<T> {
    const paymentInfo = await parsePaymentRequired(response);
    if (!paymentInfo || paymentInfo.accepts.length === 0) {
      throw new DedApiError(
        '402 Payment Required but no payment offers received',
        402
      );
    }

    const offer = paymentInfo.accepts[0];
    const paymentHeader = await buildPaymentHeader(offer, this.signer!);

    // Retry the same request with X-PAYMENT header
    const headers: Record<string, string> = {
      'X-PAYMENT': paymentHeader,
    };

    if (options.contentType) {
      headers['Content-Type'] = options.contentType;
    }

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);

    try {
      const retryResponse = await fetch(url, {
        method: options.method,
        headers,
        body: options.rawBody ?? options.body,
        signal: controller.signal,
      });

      if (!retryResponse.ok) {
        let errorBody: ApiErrorResponse | undefined;
        try {
          errorBody = (await retryResponse.json()) as ApiErrorResponse;
        } catch {
          // Response body may not be JSON
        }
        const message =
          errorBody?.message ??
          errorBody?.errors?.[0]?.message ??
          `HTTP ${retryResponse.status}: ${retryResponse.statusText}`;
        throw new DedApiError(message, retryResponse.status, errorBody);
      }

      return (await retryResponse.json()) as T;
    } finally {
      clearTimeout(timeoutId);
    }
  }
}

File: packages/python/src/constellation_digital_evidence_sdk/network/http_client.py
"""Standalone HTTP client for the DED Ingestion API using httpx."""

from __future__ import annotations

import json
from typing import Any

import httpx

from .types import DedClientConfig


class DedApiError(Exception):
    """Error thrown by DedHttpClient on API failures."""

    def __init__(self, message: str, status: int, body: dict | None = None) -> None:
        super().__init__(message)
        self.status = status
        self.body = body


class DedHttpClient:
    """Async HTTP client for the DED Ingestion API.

    Uses ``httpx.AsyncClient`` for async HTTP operations.
    Injects the API key header on authenticated requests.
    """

    def __init__(self, config: DedClientConfig) -> None:
        self._base_url = config.base_url.rstrip("/")
        self._api_key = config.api_key
        self._client = httpx.AsyncClient(timeout=config.timeout)

    async def close(self) -> None:
        await self._client.aclose()

    async def get(self, path: str, query: dict[str, str] | None = None) -> Any:
        """GET request (authenticated with API key)."""
        url = f"{self._base_url}{path}"
        headers = {"X-Api-Key": self._api_key}
        return await self._request("GET", url, headers=headers, params=query)

    async def get_public(self, path: str, query: dict[str, str] | None = None) -> Any:
        """GET request (public, no API key)."""
        url = f"{self._base_url}{path}"
        return await self._request("GET", url, params=query)

    async def post(self, path: str, body: Any) -> Any:
        """POST request (authenticated with API key)."""
        url = f"{self._base_url}{path}"
        headers = {
            "X-Api-Key": self._api_key,
            "Content-Type": "application/json",
        }
        return await self._request(
            "POST", url, headers=headers, content=json.dumps(body)
        )

    async def _request(
        self,
        method: str,
        url: str,
        headers: dict[str, str] | None = None,
        params: dict[str, str] | None = None,
        content: str | None = None,
    ) -> Any:
        response = await self._client.request(
            method, url, headers=headers, params=params, content=content
        )

        if response.status_code >= 400:
            error_body = None
            try:
                error_body = response.json()
            except Exception:
                pass
            message = "API request failed"
            if error_body:
                message = (
                    error_body.get("message")
                    or (error_body.get("errors", [{}])[0].get("message"))
                    or f"HTTP {response.status_code}"
                )
            raise DedApiError(message, response.status_code, error_body)

        return response.json()
